op-batcher/batcher/sync_actions.go

```diff
@@ -0,0 +1,144 @@
+	youngestUnsafeBlockNum := newSyncStatus.UnsafeL2.Number
+
+	if !hasBlocks {
+		s := syncActions{
+			blocksToLoad: [2]uint64{oldestUnsafeBlockNum, youngestUnsafeBlockNum},
+		}
+		l.Info("no blocks in state", "syncActions", s)
+		return s, false
+	}
+
+	oldestBlockInStateNum := oldestBlockInState.NumberU64()
+
+	if oldestUnsafeBlockNum < oldestBlockInStateNum {
+		s := syncActions{
+			clearState:   &newSyncStatus.SafeL2.L1Origin,
+			blocksToLoad: [2]uint64{oldestUnsafeBlockNum, youngestUnsafeBlockNum},
+		}
+		l.Warn("new safe head is behind oldest block in state", "syncActions", s, "oldestBlockInState", oldestBlockInState, "newSafeBlock", newSyncStatus.SafeL2)
+		return s, false
+	}
+
+	// PART 3: checks involving all blocks in state
+	newestBlockInState := blocks[blocks.Len()-1]
+	newestBlockInStateNum := newestBlockInState.NumberU64()
+
+	numBlocksToDequeue := oldestUnsafeBlockNum - oldestBlockInStateNum
+
+	// These actions apply in multiple unhappy scenarios below, where
+	// we detect that the existing state is invalidated
+	// and we need to start over from the sequencer's oldest
+	// unsafe (and not safe) block.
+	startAfresh := syncActions{
+		clearState:   &newSyncStatus.SafeL2.L1Origin,
+		blocksToLoad: [2]uint64{oldestUnsafeBlockNum, youngestUnsafeBlockNum},
+	}
+
+	if numBlocksToDequeue > uint64(blocks.Len()) {
+		// This could happen if the batcher restarted.
+		// The sequencer may have derived the safe chain
+		// from channels sent by a previous batcher instance.
+		l.Warn("safe head above unsafe head, clearing channel manager state",
+			"unsafeBlock", eth.ToBlockID(newestBlockInState),
+			"newSafeBlock", newSyncStatus.SafeL2.Number,
+			"syncActions",
+			startAfresh)
+		return startAfresh, false
+	}
+
+	if numBlocksToDequeue > 0 && blocks[numBlocksToDequeue-1].Hash() != newSyncStatus.SafeL2.Hash {
+		l.Warn("safe chain reorg, clearing channel manager state",
+			"existingBlock", eth.ToBlockID(blocks[numBlocksToDequeue-1]),
+			"newSafeBlock", newSyncStatus.SafeL2,
+			"syncActions", startAfresh)
+		return startAfresh, false
+	}
+
+	// PART 4: checks involving channels
+	for _, ch := range channels {
+		if ch.isFullySubmitted() &&
+			!ch.isTimedOut() &&
+			newSyncStatus.CurrentL1.Number > ch.MaxInclusionBlock() &&
+			newSyncStatus.SafeL2.Number < ch.LatestL2().Number {
+			// Safe head did not make the expected progress
+			// for a fully submitted channel. This indicates
+			// that the derivation pipeline may have stalled
+			// e.g. because of Holocene strict ordering rules.
+			l.Warn("sequencer did not make expected progress",
+				"existingBlock", eth.ToBlockID(blocks[numBlocksToDequeue-1]),
+				"newSafeBlock", newSyncStatus.SafeL2,
+				"syncActions", startAfresh)
+			return startAfresh, false
+		}
+	}
+
+	// PART 5: happy path
+	numChannelsToPrune := 0
+	for _, ch := range channels {
+		if ch.LatestL2().Number > newSyncStatus.SafeL2.Number {
+			break
+		}
+		numChannelsToPrune++
+	}
```

Comment:
If I see correctly, all test cases only involve a single channel. I think it would be good to have a test case that has multiple channels where only a subset is determined to be pruned, to test correctness of this loop.